lib/elm-test.js

```diff
@@ -354,16 +45,89 @@ const parsePositiveInteger = (flag /*: string */) => (
 const parseReport = (flag /*: string */) => (
   string /*: string */
 ) /*: typeof Report.Report */ => {
-  const result = Report.parse(string);
-  switch (result.tag) {
-    case 'Ok':
-      return result.value;
-    case 'Error':
-      console.error(`error: option '${flag}' ${result.error}`);
-      throw process.exit(1);
+  try {
+    return Report.parse(string);
+  } catch (error) {
+    console.error(`error: option '${flag}' ${error.message}`);
+    throw process.exit(1);
   }
 };
 
+function findClosest(
+  dir /*: string */,
+  dirToPath /*: (dir: string) => string */
+) /*: string | void */ {
+  const entry = dirToPath(dir);
+  return fs.existsSync(entry)
+    ? entry
+    : dir === path.parse(dir).root
+    ? undefined
+    : findClosest(path.dirname(dir), dirToPath);
+}
+
+function getProjectRootDir(subcommand /*: string */) {
+  const elmJsonPath = findClosest(process.cwd(), ElmJson.getPath);
+  if (elmJsonPath === undefined) {
+    const command =
+      subcommand === 'tests' ? 'elm-test' : `elm-test ${subcommand}`;
+    console.error(
+      `\`${command}\` requires an elm.json up the directory tree, but none could be found! To make one: elm init`
+    );
+    throw process.exit(1);
+  }
+  return path.dirname(elmJsonPath);
+}
+
+function getProject(subcommand /*: string */) {
+  try {
+    return Project.init(getProjectRootDir(subcommand), packageInfo.version);
+  } catch (error) {
+    console.error(error.message);
+    throw process.exit(1);
+  }
+}
+
+function getPathToElmBinary(compiler /*: string | void */) {
+  const name = compiler === undefined ? 'elm' : compiler;
+  try {
+    return path.resolve(which.sync(name));
+  } catch (_error) {
+    throw new Error(
+      compiler === undefined
+        ? `Cannot find elm executable, make sure it is installed.
+(If elm is not on your path or is called something different the --compiler flag might help.)`
+        : `The elm executable passed to --compiler must exist and be exectuble. Got: ${compiler}`
+    );
+  }
+}
+
+// Unfortunately commander is very permissive about extra arguments. Therefore,
+// we manually check for excessive arguments.
+// See: https://github.com/tj/commander.js/issues/1268
+function handleTooManyArgs(action) {
```

Comment:
nice, this and getPathToElmBinary not being top level makes everything so much nicer